lib/Doctrine/ODM/MongoDB/Iterator/BaseIterator.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ODM\MongoDB\Iterator;
+
+use Generator;
+use Traversable;
+
+/**
+ * Iterator for wrapping a Traversable and caching its results.
+ *
+ * By caching results, this iterators allows a Traversable to be counted and
+ * rewound multiple times, even if the wrapped object does not natively support
+ * those operations (e.g. MongoDB\Driver\Cursor).
+ *
+ * @internal
+ */
+final class BaseIterator implements Iterator
+{
+    private $key;
+    
+    private $current;
+
+    /** @var Generator */
+    private $iterator;
+
+    /** @var bool */
+    private $iteratorAdvanced = false;
+
+    /**
+     * Initialize the iterator and stores the first item in the cache. This
+     * effectively rewinds the Traversable and the wrapping Generator, which
+     * will execute up to its first yield statement. Additionally, this mimics
+     * behavior of the SPL iterators and allows users to omit an explicit call
+     * to rewind() before using the other methods.
+     */
+    public function __construct(Traversable $iterator)
+    {
+        $this->iterator = $this->wrapTraversable($iterator);
+        $this->storeCurrentItem();
+    }
+
+    public function toArray() : array
+    {
+        if ($this->iteratorAdvanced) {
```

Comment:
Suggest extracting this to a method to avoid duplicating logic here and in `rewind`:
```suggestion
        $this->preventRewinding(__METHOD__);
```